pkg/kubelet/status/status_manager.go

```diff
@@ -519,31 +523,30 @@ func (m *manager) syncBatch() {
 				syncedUID = mirrorUID
 			}
 			if m.needsUpdate(types.UID(syncedUID), status) {
-				updatedStatuses = append(updatedStatuses, podStatusSyncRequest{uid, status})
+				updatedStatuses = append(updatedStatuses, podStatusSyncRequest{uid, status, true})
 			} else if m.needsReconcile(uid, status.status) {
 				// Delete the apiStatusVersions here to force an update on the pod status
 				// In most cases the deleted apiStatusVersions here should be filled
 				// soon after the following syncPod() [If the syncPod() sync an update
 				// successfully].
 				delete(m.apiStatusVersions, syncedUID)
-				updatedStatuses = append(updatedStatuses, podStatusSyncRequest{uid, status})
+				updatedStatuses = append(updatedStatuses, podStatusSyncRequest{uid, status, false})
 			}
 		}
 	}()
 
 	for _, update := range updatedStatuses {
 		klog.V(5).Infof("Status Manager: syncPod in syncbatch. pod UID: %q", update.podUID)
-		m.syncPod(update.podUID, update.status)
+		m.syncPod(update.podUID, update.status, update.checkNeedsUpdate)
 	}
 }
 
 // syncPod syncs the given status with the API server. The caller must not hold the lock.
-func (m *manager) syncPod(uid types.UID, status versionedPodStatus) {
-	if !m.needsUpdate(uid, status) {
+func (m *manager) syncPod(uid types.UID, status versionedPodStatus, checkNeedsUpdate bool) {
```

Comment:
`m.needsUpdate` should be hoisted up and out of syncPod rather than passing in a boolean.

Reply:
needsUpdate is not a parameter of the func.

If you meant checkNeedsUpdate, I think it should be a parameter instead of a field.

I am open to better naming of the parameter.

Reply:
Would it be less confusing if you flipped the boolean around? I.e. `forceSync`, `alwaysSync`?

I could be wrong, but I think @smarterclayton is suggesting that rather than hiding the fact that we might skip the sync, pull the `m.needsUpdate(uid, status)` out of `syncPod` and do that check before calling `syncPod`. I agree that it's more convenient to not add `if (m.needsUpdate(uid, status)) { m.syncPod(update.podUID, update.status) }` everywhere we call `syncPod`, but it doesn't make `syncPod` cleaner and get rid of a variable that's always going to be a little confusing.

Reply:
Extracting the check to ahead of syncPod() calls seems to be bigger change.
Several places need to be updated, e.g.
```
pkg/kubelet/pod_workers.go:
                     err = p.syncPodFn(syncPodOptions{
```
If possible, I want to leave such refactoring to subsequent PRs.

I am open to renaming checkNeedsUpdate, once more people chime in and we get consensus on the final variable name.

thanks

Reply:
IMO, that's refactoring that nobody's ever likely to do, so we probably just need consensus here.

As for the several places that need updating, I don't believe it would impact anywhere else. `syncPod` is private, so it's not called outside of this file. The place that you reference in `pkg/kubelet/pod_workers.go` seems to trace back to [this method](https://github.com/kubernetes/kubernetes/blob/dd466bccde8176bd390fcf712c0752ae94444742/pkg/kubelet/kubelet.go#L1398), which calls StatusManager methods that take care of that.

Reply:
It seems to me that leaving the needsUpdate() call inside syncPod() would result in slightly shorter code.
In the future, if one more call to syncPod is added somewhere, the developer wouldn't forget that needsUpdate() may be involved.

Reply:
Shorter code is something to strive for, but doesn't take precedence over readability. `needsUpdate()` is an optimization, if I understand it correctly, so the risk that someone doesn't know to check `needsUpdate()` seems low and they'll have the other uses as examples.

Reply:
I am flexible w.r.t. how to refactor the check.

I would prefer confirmation from maintainers that the fix itself is correct.

Reply:
In general in the refactorings I was doing syncPod is already wrong, and passing a boolean to do an if check to a function is usually wrong.  Do the check above the function call, because syncPod is already supposed to be called in limited spots.

Reply:
@tedyu You have time to update this PR?

Reply:
@tedyu Let me know if you don't have the bandwidth to make the requested changes and get this over the line. If not, I can try to do it.